evm/chains/sharding/mainchain_handler/vmc_handler.py

```diff
@@ -0,0 +1,263 @@
+            sender_addr = self.primary_addr
+        return self.call_vmc('get_shard_list', [valcode_addr], sender_addr=sender_addr)
+
+    def add_header(self, header, sender_addr=None, gas=TX_GAS, gas_price=GASPRICE):
+        '''add_header(header: bytes <= 4096) -> bool
+        '''
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.send_vmc_tx(
+            'add_header',
+            [header],
+            sender_addr=sender_addr,
+            gas=gas,
+            gas_price=gas_price,
+        )
+
+    def get_period_start_prevhash(self, expected_period_number, sender_addr=None):
+        '''get_period_start_prevhash(expected_period_number: num) -> bytes32
+        '''
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.call_vmc(
+            'get_period_start_prevhash',
+            [expected_period_number],
+            sender_addr=sender_addr,
+        )
+
+    def tx_to_shard(
+            self,
+            to,
+            shard_id,
+            tx_startgas,
+            tx_gasprice,
+            data,
+            value,
+            sender_addr=None,
+            gas=TX_GAS,
+            gas_price=GASPRICE):
+        '''tx_to_shard(
+            to: address, shard_id: num, tx_startgas: num, tx_gasprice: num, data: bytes <= 4096
+           ) -> num
+        '''
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.send_vmc_tx(
+            'tx_to_shard',
+            [to, shard_id, tx_startgas, tx_gasprice, data],
+            sender_addr=sender_addr,
+            gas=gas,
+            gas_price=gas_price,
+            value=value,
+        )
+
+    def get_collation_gas_limit(self, sender_addr=None):
+        '''get_collation_gas_limit() -> num
+        '''
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.call_vmc('get_collation_gas_limit', [], sender_addr=sender_addr)
+
+    def get_collation_header_score(self, shard_id, collation_header_hash, sender_addr=None):
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.call_vmc(
+            'get_collation_headers__score',
+            [shard_id, collation_header_hash],
+            sender_addr=sender_addr,
+        )
+
+    def get_num_validators(self, sender_addr=None):
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.call_vmc('get_num_validators', [], sender_addr=sender_addr)
+
+    def get_receipt_value(self, receipt_id, sender_addr=None):
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        return self.call_vmc('get_receipts__value', [receipt_id], sender_addr=sender_addr)
+
+    # utils #######################################################
+
+    def is_vmc_deployed(self):
+        return (
+            # self.chain_handler.get_code(self._vmc_addr) != b'' and \
```

Comment:
Sorry for that. This code should not be commented out. I did it because for now I haven't finished the `get_code` in `TesterChainHandler`, which calls `PyEVMBackend.get_code`. But that part hasn't been implemented yet. I think it should be fixed when I changed to use `web3.py` tester? I will check for it. :)